src/overlay/Peer.cpp

```diff
@@ -2205,4 +2127,60 @@ Peer::fulfillDemand(FloodDemand const& dmd)
     }
 }
 
+bool
+Peer::peerKnowsHash(Hash const& hash)
+{
+    if (mTxPullMode)
+    {
+        return mTxPullMode->seenAdvert(hash);
+    }
+    return false;
+}
+
+void
+Peer::sendAdvert(Hash const& hash)
+{
+    if (!mTxPullMode)
+    {
+        throw std::runtime_error("Pull mode is not set");
+    }
+    mTxPullMode->queueOutgoingAdvert(hash);
+}
+
+void
+Peer::retryAdvert(std::list<Hash>& hashes)
+{
+    if (mTxPullMode)
+    {
+        mTxPullMode->retryIncomingAdvert(hashes);
+    }
+}
+
+bool
+Peer::hasAdvert()
+{
+    if (mTxPullMode)
+    {
+        return mTxPullMode->size() > 0;
+    }
+
+    return false;
+}
+
+std::pair<Hash, std::optional<VirtualClock::time_point>>
+Peer::popAdvert()
+{
+    if (!hasAdvert())
+    {
+        throw std::runtime_error("No advert to pop");
+    }
+
+    if (!mTxPullMode)
+    {
+        throw std::runtime_error("Pull mode is not set");
+    }
```

Comment:
I think the order of these conditionals needs to be reversed. `hasAdvert` returns false when `!mTxPullMode`, so the exception will always contain `No advert to pop` regardless of whether `popAdvert` fails because pull mode is disabled or it is enabled but empty.

Reply:
Good catch, fixed